### PR TITLE
[consensus] Introduce ExecutedBlock to refactor block_tree

### PR DESCRIPTION
**Pull request by lightmark**
## Motivation

Introduce `ExecutedBlock` which encapsulates a block and its executed result and children hash pointers.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

CI

### CODE REVIEW

File: consensus/src/chained_bft/block_storage/block_tree.rs [hunk 1]
```diff
@@ -34,26 +33,16 @@ use types::crypto_proxies::LedgerInfoWithSignatures;
...
-    /// the tree (the initial root node might not have it, because it's been already
-    /// committed). The execution results are not persisted: they're recalculated again for the
-    /// pending blocks upon restart.
-    id_to_compute_result: HashMap<HashValue, Arc<StateComputeResult>>,
+    id_to_block: HashMap<HashValue, ExecutedBlock<T>>,
     /// Root of the tree.
-    root: Arc<Block<T>>,
-    /// A certified block with highest round
-    highest_certified_block: Arc<Block<T>>,
+    root_id: HashValue,
```
Comment (zekun000) [inline consensus/src/chained_bft/block_storage/block_tree.rs]: seems to me replace Arc with id adds one more indirection, what's your reasoning behind? avoid having Block in the tree?

File: consensus/src/chained_bft/block_storage/block_tree.rs [hunk 2]
```diff
@@ -118,43 +106,50 @@ where
...
+    fn get_block(&self, block_id: &HashValue) -> &ExecutedBlock<T> {
+        self.try_get_block(block_id)
+            .expect("Block doesn't exist. Use try_get_block if None is a possible return value")
+    }
+
+    pub(super) fn try_get_block(&self, block_id: &HashValue) -> Option<&ExecutedBlock<T>> {
+        self.id_to_block.get(block_id)
+    }
+
+    pub(super) fn try_get_block_mut(
```
Comment (zekun000) [inline consensus/src/chained_bft/block_storage/block_tree.rs]: private function?

File: consensus/src/chained_bft/block_storage/block_tree.rs [hunk 3]
```diff
@@ -165,56 +160,45 @@ where
...
-
-                Ok(previous_block.clone())
-            }
-            _ => {
-                let children = match self.id_to_child.entry(block.parent_id()) {
-                    Vacant(entry) => entry.insert(Vec::new()),
-                    Occupied(entry) => entry.into_mut(),
-                };
-                children.push(block.clone());
-                counters::NUM_BLOCKS_IN_TREE.inc();
```
Comment (zekun000) [inline consensus/src/chained_bft/block_storage/block_tree.rs]: where does the counter go?

File: consensus/src/chained_bft/block_storage/block_tree.rs [hunk 4]
```diff
@@ -351,12 +330,7 @@ where
         mut newly_pruned_blocks: VecDeque<HashValue>,
     ) {
         // Update the next root
-        self.root = self
-            .id_to_block
-            .get(&root_id)
-            .expect("next_root_id must exist")
-            .clone();
-
+        self.root_id = root_id;
```
Comment (zekun000) [inline consensus/src/chained_bft/block_storage/block_tree.rs]: assert next root exist?

File: consensus/src/chained_bft/consensus_types/block.rs [hunk 5]
```diff
@@ -106,6 +108,30 @@ pub struct Block<T> {
...
+    /// The set of children for cascading pruning. Note: a block may have multiple children.
+    children: HashSet<HashValue>,
+    /// The state compute results is calculated for all the pending blocks prior to insertion to
+    /// the tree (the initial root node might not have it, because it's been already
+    /// committed). The execution results are not persisted: they're recalculated again for the
+    /// pending blocks upon restart.
+    compute_result: Arc<StateComputeResult>,
+}
+
+impl<T> Deref for ExecutedBlock<T> {
```
Comment (zekun000) [inline consensus/src/chained_bft/consensus_types/block.rs]: this is not encouraged to use other than a pointer. where do you need this?

File: consensus/src/chained_bft/block_storage/block_tree.rs [hunk 6]
```diff
@@ -118,43 +106,50 @@ where
...
         // Remove the block from the store
         self.id_to_block.remove(&block_id);
-        self.id_to_compute_result.remove(&block_id);
         self.id_to_votes.remove(&block_id);
         self.id_to_quorum_cert.remove(&block_id);
     }
 
-    pub(super) fn block_exists(&self, block_id: HashValue) -> bool {
-        self.id_to_block.contains_key(&block_id)
+    pub(super) fn block_exists(&self, block_id: &HashValue) -> bool {
```
[the remark this answers is not in the document]
Comment (lightmark) [reply consensus/src/chained_bft/block_storage/block_tree.rs]: Yes, but all the collections have api like `get(&Key)`.

File: consensus/src/chained_bft/block_storage/block_tree.rs (hunk 1 above)
Comment (lightmark) [reply consensus/src/chained_bft/block_storage/block_tree.rs]: I don't have a strong opinion here with `root_id` or `root_block`.

File: consensus/src/chained_bft/block_storage/block_tree.rs (hunk 3 above)
Comment (lightmark) [reply consensus/src/chained_bft/block_storage/block_tree.rs]: will add it back

File: consensus/src/chained_bft/consensus_types/block.rs (hunk 5 above)
Comment (lightmark) [reply consensus/src/chained_bft/consensus_types/block.rs]: Yep, maybe I should remove it.

**Pull request merged**